core/src/main/java/org/mule/processor/AbstractRequestResponseMessageProcessor.java

```diff
@@ -0,0 +1,150 @@
+        if (isNonBlocking(event))
+        {
+            return processNonBlocking(event);
+        }
+        else
+        {
+            return processBlocking(event);
+        }
+    }
+
+    protected MuleEvent processBlocking(MuleEvent event) throws MuleException
+    {
+        MessagingException exception = null;
+        try
+        {
+            return processResponse(processNext(processRequest(event)));
+        }
+        catch (MessagingException e)
+        {
+            exception = e;
+            throw e;
+        }
+        finally
+        {
+            processFinallly(event, exception);
+        }
+    }
+
+    protected MuleEvent processNonBlocking(MuleEvent event) throws MuleException
+    {
+        final ReplyToHandler originalReplyToHandler = event.getReplyToHandler();
+        event = new DefaultMuleEvent(event, new CompletionHandlerReplyToHandlerAdaptor(new CompletionHandler<MuleEvent, MessagingException>()
+        {
+            @Override
+            public void onCompletion(MuleEvent event)
+            {
+                try
+                {
+                    MuleEvent response = processResponse(new DefaultMuleEvent(event, originalReplyToHandler));
+                    if (!NonBlockingVoidMuleEvent.getInstance().equals(response))
+                    {
+                        originalReplyToHandler.processReplyTo(response, null, null);
+                    }
+                    processFinallly(event, null);
+                }
+                catch (MuleException e)
+                {
+                    onFailure(new MessagingException(event, e));
+                }
+            }
+
+            @Override
+            public void onFailure(MessagingException exception)
+            {
+                originalReplyToHandler.processExceptionReplyTo(exception.getEvent(), exception, null);
+                processFinallly(exception.getEvent(), exception);
+            }
+        }));
+
+        try
+        {
+            MuleEvent result = processNext(processRequest(event));
+            if (!(result instanceof NonBlockingVoidMuleEvent))
+            {
+                return processResponse(event);
+            }
+            else
+            {
+                return result;
+            }
+        }
+        catch (MessagingException exception)
+        {
+            processFinallly(event, exception);
+            throw exception;
+        }
+    }
+
+    private boolean isNonBlocking(MuleEvent event)
+    {
+        return event.isAllowNonBlocking() && event.getReplyToHandler() != null && next != null;
+    }
+
+    /**
+     * Processes the request phase before the next message processor is invoked.
+     *
+     * @param event event to be processed.
+     * @return result of request processing.
+     * @throws MuleException exception thrown by implementations of this method whiile performing response processing
+     */
+    protected MuleEvent processRequest(MuleEvent event) throws MuleException
```

Comment:
I disagree.  Method is performing any request phase processing. i.e. anything that in an InterceptingMessageProcessor comes before processNext()